integration_tests/src/main/python/parquet_pyarrow_test.py

```diff
@@ -0,0 +1,255 @@
+    """Verify that the result from the pyarrow(already adapted to CPU data format) and the GPU are equal"""
+    try:
+        _assert_equal(cpu_adapted_from_pyarrow, gpu, float_check=get_float_check(), path=[],
+                      cpu_data_adapted_from="pyarrow")
+    except:
+        sys.stdout.writelines(difflib.unified_diff(
+            a=[f"{x}\n" for x in cpu_adapted_from_pyarrow],
+            b=[f"{x}\n" for x in gpu],
+            fromfile='pyarrow OUTPUT',
+            tofile='GPU OUTPUT'))
+        raise
+
+
+def assert_gpu_and_pyarrow_are_compatible(base_write_path, gen_list, conf={}):
+    """
+    Test scenarios:
+      Write by pyarrow, test readings on pyarrow and GPU
+      Write by GPU, test readings on pyarrow and GPU
+    The data is collected back to the driver and compared here, so be
+    careful about the amount of data returned.
+    """
+    conf = _prep_incompat_conf(conf)
+    pyarrow_parquet_path = base_write_path + "/pyarrow"
+    gpu_parquet_path = base_write_path + "/gpu"
+
+    def write_on_pyarrow():
+        pa_table = _gen_pyarrow_table(gen_list)
+        pa_pq.write_table(pa_table, pyarrow_parquet_path)
+
+    def write_on_gpu():
+        with_gpu_session(lambda spark: gen_df(spark, gen_list).coalesce(1).write.parquet(gpu_parquet_path))
+
+    def read_on_pyarrow(parquet_path):
+        print('### pyarrow RUN ###')
+        pyarrow_start = time.time()
+        pyarrow_table = pa_pq.read_table(parquet_path)
+        # adapt pyarrow table to CPU result
+        from_pyarrow = _convert_pyarrow_table_to_cpu_object(pyarrow_table)
+        pyarrow_end = time.time()
+        return (from_pyarrow, pyarrow_end - pyarrow_start)
+
+    def read_on_gpu(parquet_path):
+        print('### GPU RUN ###')
+        gpu_start = time.time()
+        # collect all data to driver
+        from_gpu = with_gpu_session(lambda spark: spark.read.parquet(parquet_path).collect(), conf=conf)
+        gpu_end = time.time()
+        return (from_gpu, gpu_end - gpu_start)
+
+    def compare_reads(parquet_path):
+        from_pyarrow, time_pyarrow = read_on_pyarrow(parquet_path)
+        from_gpu, time_gpu = read_on_gpu(parquet_path)
+        print('### collect: GPU TOOK {} pyarrow TOOK {} ###'.format(time_pyarrow, time_gpu))
+        # sort locally
+        from_pyarrow.sort(key=_RowCmp)
+        from_gpu.sort(key=_RowCmp)
+        assert_equal(from_pyarrow, from_gpu)
+
+    # write on pyarrow, compare reads between pyarrow and GPU
+    write_on_pyarrow()
+    compare_reads(pyarrow_parquet_path)
+
+    # write on GPU, compare reads between pyarrow and GPU
+    write_on_gpu()
+    compare_reads(gpu_parquet_path)
+
+
+# types for test_parquet_read_round_trip_write_by_pyarrow
+parquet_gens_list = [
+    [boolean_gen, byte_gen, short_gen, int_gen, long_gen, float_gen, double_gen],  # basic types
+    # [date_gen], TODO: not pass
+    # [timestamp_gen], TODO: not pass
+    # [binary_gen], TODO: not pass
```

Comment:
I think binary_gen failed because it should convert to `binaryarray` instead of `bytes`. Modifying `_convert_pyarrow_column_to_cpu_object` as follows will make it pass.

```python
def _convert_pyarrow_binary_object_to_cpu_object(pyarrow_binary_scalar):
    """Convert pyarrow binary to bytearray"""
    # pyarrow.BinaryScalar obj to bytearray
    py_bytearray = pyarrow_binary_scalar.as_py()
    if py_bytearray is None:
        return None
    else:
        return bytearray(py_bytearray)


def _convert_pyarrow_column_to_cpu_object(pyarrow_column, pa_type):
    """ convert pyarrow column to a column contains CPU type objects"""
    if isinstance(pa_type, pa.StructType):
        # struct column
        return [_convert_pyarrow_struct_object_to_cpu_object(v) for v in pyarrow_column]
    elif isinstance(pa_type, pa.ListType):
        # list column
        return [_convert_pyarrow_list_object_to_cpu_object(v) for v in pyarrow_column]
    elif isinstance(pa_type, pa.MapType):
        # map column
        return [_convert_pyarrow_map_object_to_cpu_object(v) for v in pyarrow_column]
    elif isinstance(pa_type, type(pa.binary())):
        # binary column should be converted to bytearray
        return [_convert_pyarrow_binary_object_to_cpu_object(v) for v in pyarrow_column]
    else:
        # other type column
        return [v.as_py() for v in pyarrow_column]
```

Reply:
Good catch